BLE_GAP/source/main.cpp

```diff
@@ -0,0 +1,456 @@
+
+        if (error) {
+            print_error(error, "Error returned by BLE::init.\r\n");
+            return;
+        }
+
+        /* to show we're running we'll blink every 500ms */
+        _event_queue.call_every(500, this, &GAPDevice::blink);
+
+        /* this will not return until shutdown */
+        _event_queue.dispatch_forever();
+    };
+
+private:
+    /** This is called when BLE interface is initialised and starts the first mode */
+    void on_init_complete(BLE::InitializationCompleteCallbackContext *event)
+    {
+        if (event->error) {
+            print_error(event->error, "Error during the initialisation\r\n");
+            return;
+        }
+
+        /* start our first demo mode,
+         * all calls are serialised on the user thread through the event queue */
+        _event_queue.call(this, &GAPDevice::demo_mode_start);
+    };
+
+    /** Set up and start advertising */
+    void advertise()
+    {
+        ble_error_t error;
+        GapAdvertisingData advertising_data;
+
+        /* add advertising flags */
+        advertising_data.addFlags(GapAdvertisingData::LE_GENERAL_DISCOVERABLE |
+                                  GapAdvertisingData::BREDR_NOT_SUPPORTED);
+
+        /* add device name */
+        advertising_data.addData(GapAdvertisingData::COMPLETE_LOCAL_NAME,
+                                 DEVICE_NAME,
+                                 sizeof(DEVICE_NAME));
+
+        error = _ble.gap().setAdvertisingPayload(advertising_data);
+
+        if (error) {
+            print_error(error, "Error during Gap::setAdvertisingPayload\r\n");
+            return;
+        }
+
+        /* set the advertising parameters according to currently selected set,
+         * see @AdvertisingType_t for explanation of modes */
+        GapAdvertisingParams::AdvertisingType_t adv_type =
+                advertising_params[_set_index].adv_type;
+
+        /* how many milliseconds between advertisements, lower interval
+         * increases the chances of being seen at the cost of more power */
+        uint16_t interval = advertising_params[_set_index].interval;
+
+        /* advertising will continue for this many seconds or until connected */
+        uint16_t timeout = advertising_params[_set_index].timeout;
+
+        _ble.gap().setAdvertisingType(adv_type);
+        _ble.gap().setAdvertisingInterval(interval);
+        _ble.gap().setAdvertisingTimeout(timeout);
+
+        error = _ble.gap().startAdvertising();
+
+        if (error) {
+            print_error(error, "Error during Gap::startAdvertising.\r\n");
+            return;
+        }
+
+        printf("Advertising started.\r\n");
+    };
+
+    /** Set up and start scanning */
+    void scan()
+    {
+        ble_error_t error;
+
+        /* scanning happens repeatedly, interval is the number of milliseconds
+         * between each cycle of scanning */
+        uint16_t interval = scanning_params[_set_index].interval;
```

Comment:
this would be better as:
const uint16_t &interval = scanning_params[_set_index].interval;
but omitted for brevity

Reply:
I guess it depends on what the optimising compiler will do. A naive one will allocate extra bytes on the stack without the refs.

Reply:
A compiler will probably allocate bytes to store a pointer to the element referenced. If the structure and not the field was referenced then we may save stack space but it is not worth it in my opinion. Clarity is the main driver.

Reply:
I agree on clarity which is why I didn't bother with it but no pointer allocation should happen since this memory location for this const data is known at compile time and a compiler would optimise it out making the ref completely transparent and only exist in the source file but having no effect on code generation.

Reply:
`_set_index` is not known at compile time, it changes at runtime; Of course the compiler can generate code that load the data based on `_set_index` rather than store the reference on the stack but it can do the same if a value is used (see [this example](https://godbolt.org/g/jL6opg)). 

It is very hard to speculate on what the compiler will do these days.